src/System.CommandLine/Invocation/InvocationPipeline.cs

```diff
@@ -24,11 +24,32 @@ internal static async Task<int> InvokeAsync(ParseResult parseResult, Cancellatio
                 {
                     for (int i = 0; i < parseResult.NonexclusiveActions.Count; i++)
                     {
-                        await parseResult.NonexclusiveActions[i].InvokeAsync(parseResult, cts.Token);
+                        var action = parseResult.NonexclusiveActions[i];
+
+                        switch (action)
+                        {
+                            case SynchronousCliAction syncAction:
+                                syncAction.Invoke(parseResult);
+                                break;
+                            case AsynchronousCliAction asyncAction:
+                                await asyncAction.InvokeAsync(parseResult, cts.Token);
+                                break;
+                        }
                     }
                 }
 
-                Task<int> startedInvocation = parseResult.Action.InvokeAsync(parseResult, cts.Token);
+                Task<int> startedInvocation;
+                switch (parseResult.Action)
+                {
+                    case SynchronousCliAction syncAction:
+                        startedInvocation = Task.FromResult(syncAction.Invoke(parseResult));
```

Comment:
We should most probably (I am not 100% sure) call `Task.Run` here and provide it with the cancellation token if we want to get cancellation to work, otherwise we are going to:

1. perform a blocking call and wait until it's finished
2. register for cancellation and do all other things that won't matter as the job will be already finished

```suggestion
                        startedInvocation = Task.Run(() => syncAction.Invoke(parseResult), cts.Token);
```

Reply:
If we don't start task of some kind for performing the synchronous action, the code below will be pointless, as it's going to register for cancellation after work has been finished:

1. Perform blocking action call and wait until it's finished
2. Register for cancellation
3. Return immediately from `Task.WhenAny` as the work has been already done
https://github.com/dotnet/command-line-api/blob/2824b6ff68799f5951c6c15646c92c314faff8ea/src/System.CommandLine/Invocation/InvocationPipeline.cs#L68-L69

If the sync call never finishes, we won't be able to cancel it gracefully.

If for some reason we don't want to support cancellation in this scenario (I doubt), we should simple return here

```diff
- startedInvocation = Task.FromResult(syncAction.Invoke(parseResult));
+ return syncAction.Invoke(parseResult);
```

Reply:
It's up to the handler (the developer's code) to implement cancellation, not the caller (our code here). For example, someone could do this inside a synchronous handler (assuming we figure out how to provide the cancellation token):

```csharp

while (!cancellationToken.IsCancellationRequested)
{
  // do stuff
}

CleanUp();

```